File: app/office2pdf/src/office2pdf.py
import os
import sys
import subprocess
import tempfile

def main():

    # Placeholder for the actual office to pdf conversion logic
    if len(sys.argv) != 2:
        print("Usage: office2pdf <PATH>")
        sys.exit(1)

    path = sys.argv[1]

    file_name = path.split('/')[-1]
    file_name_extension = file_name.rsplit('.', 1)[-1].lower()

    if file_name_extension not in ['ppt', 'pptx', 'doc', 'docx', 'xls', 'xlsx']:
        print(f"Unsupported file type: {file_name_extension}")
        sys.exit(1)

    tmp_pdf_path = convert_office_to_pdf(path)


def convert_office_to_pdf(office_file_path):

    try:
        output_dir = os.path.dirname(office_file_path)

        # Create a temporary user directory for LibreOffice
        with tempfile.TemporaryDirectory() as temp_user_dir:
            env = os.environ.copy()
            env['HOME'] = temp_user_dir
            env['TMPDIR'] = temp_user_dir
            env['LANG'] = 'ja_JP.UTF-8'
            env['LC_ALL'] = 'ja_JP.UTF-8'
            env['JAVA_TOOL_OPTIONS'] = '-Djava.awt.headless=true'

            cmd = [
                'libreoffice',
                '--headless',
                '--invisible',
                '--nodefault',
                '--nolockcheck',
                '--nologo',
                '--norestore',
                '--convert-to', 'pdf',
                '--outdir', output_dir,
                f'-env:UserInstallation=file://{temp_user_dir}/.config/libreoffice',
                office_file_path,
            ]

            result = subprocess.run(cmd, capture_output=True, text=True, timeout=60, env=env)

            if result.stdout:
                print(result.stdout, end='')
            if result.stderr:
                print(result.stderr, end='')

            if result.returncode == 0:
                base_name = os.path.splitext(os.path.basename(office_file_path))[0]
                pdf_path = os.path.join(output_dir, f"{base_name}.pdf")
                if not os.path.exists(pdf_path):
                    print(f"Error: PDF file not created at {pdf_path}")
                    return None
                #print(f"Successfully converted to: {pdf_path}")
                return pdf_path
            else:
                print(f"LibreOffice conversion failed: {result.stderr}")
                return None

    except subprocess.TimeoutExpired:
        print("Error: Conversion timed out after 60 seconds")
        return None
    except FileNotFoundError:
        print("Error: LibreOffice not found. Please install LibreOffice.")
        return None
    except Exception as e:
        print(f"Error during conversion: {e}")
        return None


main()

File: app/office2pdf/src/office2pdf.pl
#!/usr/bin/perl
use strict;
use warnings;
use File::Basename;
use File::Temp qw(tempdir);
use IPC::Run3;

sub main {
    if (@ARGV != 2) {
        print "Usage: office2pdf <INPUT_PATH> <OUTPUT_PATH>\n";
        exit 1;
    }

    my ($path, $output_path) = @ARGV;
    
    my ($name, $dir, $suffix) = fileparse($path, qr/\.[^.]*/);
    my $extension = lc($suffix);
    $extension =~ s/^\.//;  # Remove leading dot
    
    unless ($extension =~ /^(ppt|pptx|doc|docx|xls|xlsx)$/) {
        print "Unsupported file type: $extension\n";
        exit 1;
    }
    
    convert_office_to_pdf($path, $output_path);
}

sub convert_office_to_pdf {
    my ($office_file_path, $desired_output_path) = @_;
    
    my $temp_output_dir = dirname($office_file_path);
    
    # Create a temporary user directory for LibreOffice
    my $temp_user_dir = tempdir(CLEANUP => 1);
    
    # Set up environment
    local %ENV = %ENV;
    $ENV{HOME} = $temp_user_dir;
    $ENV{TMPDIR} = $temp_user_dir;
    $ENV{LANG} = 'ja_JP.UTF-8';
    $ENV{LC_ALL} = 'ja_JP.UTF-8';
    $ENV{JAVA_TOOL_OPTIONS} = '-Djava.awt.headless=true';
    
    my @cmd = (
        'libreoffice',
        '--headless',
        '--invisible',
        '--nodefault',
        '--nolockcheck',
        '--nologo',
        '--norestore',
        '--convert-to', 'pdf',
        '--outdir', $temp_output_dir,
        "-env:UserInstallation=file://$temp_user_dir/.config/libreoffice",
        $office_file_path,
    );
    
    my ($stdout, $stderr);
    eval {
        local $SIG{ALRM} = sub { die "timeout\n" };
        alarm 60;  # 60 second timeout
        
        run3(\@cmd, \undef, \$stdout, \$stderr);
        my $exit_code = $? >> 8;
        
        alarm 0;
        
        print $stdout if $stdout;
        print STDERR $stderr if $stderr;
        
        if ($exit_code == 0) {
            my $base_name = basename($office_file_path);
            $base_name =~ s/\.[^.]*$//;  # Remove extension
            my $temp_pdf_path = "$temp_output_dir/$base_name.pdf";
            
            unless (-f $temp_pdf_path) {
                print "Error: PDF file not created at $temp_pdf_path\n";
                return;
            }
            
            # Move the PDF to the desired output location
            rename($temp_pdf_path, $desired_output_path) or do {
                print "Error: Could not move PDF to $desired_output_path: $!\n";
                return;
            };
            
            return $desired_output_path;
        } else {
            print "LibreOffice conversion failed: $stderr\n" if $stderr;
            return;
        }
    };
    
    if ($@) {
        if ($@ =~ /timeout/) {
            print "Error: Conversion timed out after 60 seconds\n";
        } elsif ($@ =~ /failed to start/) {
            print "Error: LibreOffice not found. Please install LibreOffice.\n";
        } else {
            print "Error during conversion: $@\n";
        }
        return;
    }
}

main();